### DISCUSSION TITLE
Log improvements

### DISCUSSION THREAD

**Discussion by aksiome** [Gunivers/Bookshelf#95] [Polls]
Currently the log module uses tags. While this is a nice way to control logs for each player it also brings some problem. For example when working with the lib you expect to have at least the error log enabled so that when you use a function in a wrong way you get a message of what went wrong. This is especially true in the macro world. This force to add the bs.log._.error tag to each new player which i think is not very conveniant. The only solution i can think of is to use scores that could look like this:
`$log._.level bs.data` | `$log.<feature>.level bs.data` 

Issues:
- less explicit than `.debug`, `.warn`, ... since it's a numeric level
- we loose the ability to control logs for each player

Advantages:
- we avoid the tag mess where people first set their tag to error then info then something else on an other feature. This can quickly lead to a nightmare for the user to maintain.
- we gain the ability to control logs for all players.

**Comment by aksiome** [comment]
I also forgot to mention that this is not too complex to implement since we can use predicates this would also imrprove a little bit performances. I'll be happy to write them myself ^^

**Comment by VForiel** [comment]
Datapackers often work alone or on a local world. It will then not change anything for most of them. Also, the tag mess is still managed by the Bookshelf menu, it shoudl'nt be managed directly using commands.

However, I think it's a good idea to prevent people who will print their tags and be afraid of what they will see, complain on the lib etc. Unfortunately, datapackers aren't always the furthest slithering penguins ^^'

**Comment by theogiraudet** [comment]
I don't really like scores to manage something like that. They lack of expressivity and I think it's important in our case.
I agree about the number of tags, we can propose a function to cleanup all logs tag, it will be pretty easy to develop.
We can also give by default the tag `bs.log._.error` to each person considered as a developper.

**Comment by aksiome** [comment]
I honestly have no preferences over one or the other, I find both solutions to be a bit janky.
The only thing that really matter imo is that you get an error message without having to setup the tag. That's why i proposed scores since as Leirof said datapatckers often work alone. But some would argue that for a public server you need to be able to log only to a certain set of people. I'm not completly against setting the tag on login but it's hard to determine who should get it.

In my head having both could work like this:
When developping i use the score (they work gloabally on all features)
When testing a map i use tags to avoid spamming others (since it's very situational i'm not worried as much about the tag mess)

An other solution I can think of to keep only the tags is to be able to control them gloabally:
`function #bs.log:set_level {level:"warn",feature:"*"}` (will set to all players, and those who join the game after)
`function #bs.log:set_player_level {level:"warn",feature:"*"}`  (will set only for the current player)

I'm just worried about getting the list of features in order to remove the tags, especially if people use this module with their own custom features. An other problem is that we need to test for login or first join. Both I think are not ideal for a library, even through an advancement...

Also we could use both scores and tags like i mentioned earlier but hide how they are implemented through helper functions (like set_level and set_player_level), this remove the need for a login check.

EDIT:
Functions that I named set_level and set_player_level are not really needed if everything can be managed by the bs menu. But they can still be nice to have!

**Comment by VForiel** [comment]
An important thing to take in account is that when the mapmaker will publish the map, he will probably remove all player data files and scoreboard files. Having the logs by default imply to potentially create undesired error messages on published maps

However, as all the logs are manageable via a menu, we can simply imagine an highlighted button on the first debug page which enable by default, and for all users, the error or the warn level of debug

**Reply by aksiome** [reply]
Yeah you are right i was thinking that deleting the score would just default to no log but then the score is not set by default x), Im a bit stupid...

Idk how to stay simple and clear on the menu since im not really familiar with it but I think it's really important. When i first did some test with validation i found it weird to not get any message by default on error.

Maybe we need something like an env mode variable for prod, or dev / debug?

**Reply by VForiel** [reply]
> i found it weird to not get any message by default on error

Maybe because logs are not implemented haha (not the module itself, but all the debug lines in the different features ^^)

**Comment by aksiome** [comment]
Im closing this since I think we have our best answer with theo suggestion, keep tags, add a cleanup, and put the warn or error tag by defaut when the player is recognized as a dev (still need to determine how he is recognized as such :p) Also add a button to enable log for all features on the menu!
